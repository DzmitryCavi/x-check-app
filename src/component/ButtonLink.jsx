/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import { Button } from 'antd'

const ButtonAntd = ({ linkTo, linkMode, match, location, history, staticContext, ...props }) =>
  linkTo ? (
    <Button
      {...props}
      onClick={() => {
        if (linkMode === 'replace') {
          history.replace(linkTo)
        } else {
          history.push(linkTo)
        }
      }}
    />
  ) : (
    <Button {...props} />
  )

ButtonAntd.propTypes = {
  linkTo: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  linkMode: PropTypes.oneOf(['replace']),
  location: PropTypes.instanceOf(Object).isRequired,
  match: PropTypes.instanceOf(Object).isRequired,
  history: PropTypes.instanceOf(Object).isRequired,
  staticContext: PropTypes.instanceOf(Object),
}

ButtonAntd.defaultProps = {
  linkTo: null,
  linkMode: null,
  staticContext: null,
}

export default withRouter(ButtonAntd)
