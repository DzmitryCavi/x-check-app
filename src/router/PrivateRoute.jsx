/* eslint-disable react/jsx-props-no-spreading */
import React, { useEffect } from 'react'
import { Redirect, Route } from 'react-router-dom'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { notification } from 'antd'

const PrivateRoute = ({ component: Component, isLoggedIn, user, allowedRoles, ...rest }) => {
  const ResultComponent = (props) => {
    const isAuth = isLoggedIn && user
    const isAccess = (isAuth && allowedRoles.includes(user.role)) || !allowedRoles.length

    useEffect(() => {
      if (isAuth && !isAccess) {
        notification.error({
          className: 'app-notification app-notification--error',
          message: 'Error',
          description: 'What are you doing... You do not have access to this page!',
        })
      }
    }, [isAuth, isAccess])

    if (!isAuth) return <Redirect to="/login" />
    if (!isAccess) return <Redirect to="/" />
    return <Component {...props} />
  }

  return <Route {...rest} component={ResultComponent} />
}

PrivateRoute.defaultProps = {
  user: null,
  allowedRoles: [],
}

PrivateRoute.propTypes = {
  isLoggedIn: PropTypes.bool.isRequired,
  user: PropTypes.oneOfType([PropTypes.object]),
  allowedRoles: PropTypes.arrayOf(PropTypes.string),
  component: PropTypes.instanceOf(Object).isRequired,
}

const mapStateToProps = (state) => {
  return {
    isLoggedIn: state.auth.isLoggedIn,
    user: state.auth.user,
  }
}

export default connect(mapStateToProps, null)(PrivateRoute)
