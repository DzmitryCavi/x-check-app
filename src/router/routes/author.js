import TasksList from '../../page/author/tasks/TasksList'
import TaskCreate from '../../page/author/tasks/TaskCreate'
import TaskEdit from '../../page/author/tasks/TaskEdit'
import TaskView from '../../page/author/tasks/TaskView'
import CategoryCreate from '../../page/author/categories/CategoryCreate'
import CategoryEdit from '../../page/author/categories/CategoryEdit'

import { authorRoutes as routes } from '.'

export default [
  {
    path: routes.categories.create,
    component: CategoryCreate,
    breadcrumb: 'Create Category',
  },
  {
    path: routes.tasks.view,
    component: TaskView,
    breadcrumb: 'View Task',
  },
  {
    path: routes.tasks.create,
    component: TaskCreate,
    breadcrumb: 'Create Task',
  },
  {
    path: routes.tasks.edit,
    component: TaskEdit,
    breadcrumb: 'Edit Task',
  },
  {
    path: routes.tasks.list,
    component: TasksList,
    breadcrumb: 'Tasks',
  },
  {
    path: routes.categories.edit,
    component: CategoryEdit,
    breadcrumb: 'Edit Category',
  },
]
