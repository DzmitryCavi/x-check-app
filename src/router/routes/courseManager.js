import { ApartmentOutlined, UnorderedListOutlined } from '@ant-design/icons'

import TasksList from '../../page/courseManager/tasks/TasksList'
import TaskView from '../../page/courseManager/tasks/TaskView'
import MarksList from '../../page/courseManager/marks/MarksList'
import MarkInfo from '../../page/courseManager/marks/MarkInfo'

import { courseManagerRoutes as routes } from '.'
import Dispute from '../../page/Dispute'

export default [
  {
    path: routes.tasks.view,
    component: TaskView,
    breadcrumb: 'View Task',
  },
  {
    path: routes.tasks.list,
    component: TasksList,
    breadcrumb: 'Tasks List',
    description: 'Tasks management',
    navigation: {
      label: 'Tasks List',
      icon: ApartmentOutlined,
      color: '#52c41a',
    },
  },
  {
    path: routes.reviews.info,
    component: MarkInfo,
    breadcrumb: 'Mark Info',
  },
  {
    path: routes.reviews.dispute,
    component: Dispute,
    breadcrumb: 'Dispute',
    description: 'Dispute managment',
  },
  {
    path: routes.reviews.list,
    component: MarksList,
    breadcrumb: 'Marks List',
    description: 'Marks management',
    navigation: {
      label: 'Marks List',
      icon: UnorderedListOutlined,
      color: '#ffaf1c',
    },
  },
]
