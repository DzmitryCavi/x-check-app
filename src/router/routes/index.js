export const publicRoutes = {
  home: '/',
  login: '/login',
  notFound: '*',
}

export const authorRoutes = {
  tasks: {
    list: '/author/tasks',
    create: '/author/tasks/create',
    edit: '/author/tasks/:taskId/edit',
    view: '/author/tasks/:taskId/view',
  },
  categories: {
    create: '/author/tasks/:taskId/categories/create',
    edit: '/author/categories/:categoryId/edit',
  },
}

export const studentRoutes = {}

export const supervisorRoutes = {
  requests: {
    list: '/supervisor/requests',
    review: '/supervisor/requests/:requestId/review',
  },
}

// ...
