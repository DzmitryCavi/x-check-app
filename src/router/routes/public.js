import Home from '../../page/Home'
import Login from '../../page/Login'
import NotFound from '../../page/NotFound'

import EmptyLayout from '../../layouts/Empty'

import { publicRoutes as routes } from '.'

export default [
  {
    path: routes.home,
    component: Home,
    breadcrumb: 'Home',
    exact: true,
  },
  {
    path: routes.login,
    component: Login,
    layout: EmptyLayout,
    exact: true,
  },
  {
    path: routes.notFound,
    component: NotFound,
    exact: true,
  },
]
