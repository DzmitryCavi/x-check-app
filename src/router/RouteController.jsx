/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'
import { Route } from 'react-router-dom'
import PropTypes from 'prop-types'

import PrivateRoute from './PrivateRoute'
import DefaultLayout from '../layouts/Default'

const RouteController = ({ component: Component, layout: Layout, allowedRoles, ...rest }) => {
  const ComponentWithLayout = (props) => (
    <Layout>
      <Component {...props} />
    </Layout>
  )

  if (allowedRoles.length)
    return <PrivateRoute {...rest} component={ComponentWithLayout} allowedRoles={allowedRoles} />
  return <Route {...rest} component={ComponentWithLayout} />
}

RouteController.defaultProps = {
  layout: DefaultLayout,
  allowedRoles: [],
}

RouteController.propTypes = {
  component: PropTypes.instanceOf(Object).isRequired,
  layout: PropTypes.instanceOf(Object),
  allowedRoles: PropTypes.arrayOf(PropTypes.string),
}

export default RouteController
