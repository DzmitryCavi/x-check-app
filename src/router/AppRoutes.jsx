/* eslint-disable react/jsx-props-no-spreading */
import React from 'react'
import { BrowserRouter as Router, Switch } from 'react-router-dom'
import RouteController from './RouteController'

import publicRoutes from './routes/public'
import authorRoutes from './routes/author'
import supervisorRoutes from './routes/supervisor'
import courseManagerRoutes from './routes/courseManager'

const AppRoutes = () => {
  return (
    <Router>
      <Switch>
        {/* Author */}
        {authorRoutes.map((route) => (
          <RouteController key={route.path} {...route} allowedRoles={['author']} />
        ))}

        {/* Supervisor */}
        {supervisorRoutes.map((route) => (
          <RouteController key={route.path} {...route} allowedRoles={['supervisor']} />
        ))}

        {/* Course Manager */}
        {courseManagerRoutes.map((route) => (
          <RouteController key={route.path} {...route} allowedRoles={['course_manager']} />
        ))}

        {/* Public */}
        {publicRoutes.map((route) => (
          <RouteController key={route.path} {...route} />
        ))}
      </Switch>
    </Router>
  )
}

export default AppRoutes
