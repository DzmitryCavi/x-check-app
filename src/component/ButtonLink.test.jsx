import React from 'react'
import { MemoryRouter, Route } from 'react-router-dom'
import { create } from 'react-test-renderer'
import ButtonLink from './ButtonLink'

import { authorRoutes } from '../router/routes'

describe('ButtonLink component', () => {
  it('Matches the snapshot', () => {
    const component = create(
      <MemoryRouter
        initialEntries={['/author', '/author/tasks', '/author/tasks/create']}
        initialIndex={1}
      >
        <Route path="/author/tasks">
          <ButtonLink linkTo={authorRoutes.tasks.create} />
        </Route>
      </MemoryRouter>,
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
