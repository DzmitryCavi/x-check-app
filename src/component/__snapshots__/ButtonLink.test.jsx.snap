// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonLink component Matches the snapshot 1`] = `
<button
  className="ant-btn"
  onClick={[Function]}
  type="button"
/>
`;

exports[`ButtonLink component atches the snapshot 1`] = `
<button
  className="ant-btn"
  onClick={[Function]}
  type="button"
/>
`;
